tensorflow/core/kernels/mkl_concat_op.cc

```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
+          new memory::desc(srcs_md[i].data));
+      context_.src_md.push_back(source_md);
+
+      std::shared_ptr<mkldnn::memory::primitive_desc> src_mpd(
+          new memory::primitive_desc(*source_md, cpu_engine_));
+      context_.src_pd_shdptr.push_back(src_mpd);
+
+      std::shared_ptr<mkldnn::memory> src_mem(
+          new mkldnn::memory(*src_mpd, DummyData));
+      context_.data_mem_shdptr.push_back(src_mem);
+    }
+
+    for (size_t i = 0; i < concat_fwd_dims.num_inputs; i++) {
+      context_.data_mem.push_back(*context_.data_mem_shdptr[i]);
+      context_.src_pd.push_back(*context_.src_pd_shdptr[i]);
+    }
+    // create a concat primitive descriptor
+    context_.fwd_pd.reset(new concat::primitive_desc(
+        concat_fwd_dims.concat_dims, context_.src_pd));
+
+    // store the expected memory format
+    context_.dst_md.reset(new memory::desc({concat_fwd_dims.dst_dims},
+                                           MklDnnType<T>(),
+                                           concat_fwd_dims.mkl_common_format));
+    context_.dst_pd.reset(
+        new memory::primitive_desc(*context_.dst_md, cpu_engine_));
+
+    // create memory primitive based on dummy data
+    context_.dst_mem.reset(new memory(*context_.dst_pd, DummyData));
+
+    // create concat primitive
+    context_.concat_fwd.reset(
+        new concat(*context_.fwd_pd, context_.data_mem, *context_.dst_mem));
+
+    context_.fwd_primitives.push_back(*context_.concat_fwd);
+    return;
+  }
+
+  struct ConcatFwdContext context_;
+  engine cpu_engine_;
+};
+
+// class to create/cache the mkl concat primitives based on the
+// input and output parameters
+template <typename T>
+class MklConcatFwdPrimitiveFactory : public MklPrimitiveFactory<T> {
+ public:
+  static MklConcatFwdPrimitive<T>* Get(
+      const MklConcatFwdParams& concat_fwd_dims,
+      const std::vector<memory::desc>& srcs_md, bool do_not_cache) {
+    MklConcatFwdPrimitive<T>* concat_fwd = nullptr;
+
+    if (do_not_cache) {
+      // Always create new primitive
+      concat_fwd = new MklConcatFwdPrimitive<T>(concat_fwd_dims, srcs_md);
+    } else {
+      // try to find a suitable one in pool
+      concat_fwd = dynamic_cast<MklConcatFwdPrimitive<T>*>(
+          MklConcatFwdPrimitiveFactory<T>::GetInstance().GetConcatFwd(
+              concat_fwd_dims));
+      if (concat_fwd == nullptr) {
+        concat_fwd = new MklConcatFwdPrimitive<T>(concat_fwd_dims, srcs_md);
+        MklConcatFwdPrimitiveFactory<T>::GetInstance().SetConcatFwd(
+            concat_fwd_dims, concat_fwd);
+      }
+    }
+
+    return concat_fwd;
+  }
+
+ private:
+  MklConcatFwdPrimitiveFactory() {}
+  ~MklConcatFwdPrimitiveFactory() {}
+
+  static MklConcatFwdPrimitiveFactory& GetInstance() {
+    static MklConcatFwdPrimitiveFactory instance_;
+    return instance_;
+  }
+
+  static string CreateKey(const MklConcatFwdParams& concat_fwd_dims) {
+    string prefix = "Concat_fwd_";
```

Comment:
Other files seem to use all lowercase letters for keys. Let's make this consistent.
```suggestion
    string prefix = "concat_fwd_";
```

Reply:
ok